content/posts/2023-12-31_constraining_complexity.markdown

```diff
@@ -0,0 +1,623 @@
+
+```go
+func F(x, y, z bool) bool {
+    return (!x || z) && (y || z) && (x || !z)
+}
+```
+
+This assumption may seem like a cheat, but importantly,
+[SAT is still `NP`-complete with it](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem#Conjunctive_normal_form_2).
+
+The first step in our reduction is to model our boolean variables. Every
+variable can be either true or false and it can appear negated or not negated.
+We encode that by defining two interfaces per variable[^4]:
+
+```go
+type X interface { X() }     // X is assigned "true"
+type NotX interface{ NotX()} // X is assigned "false"
+```
+
+This allows us to translate our formula directly, using union-elements for "or"
+and interface-embedding for "and":
+
+```go
+// Represents (!x || z) && (y || z) && (x || !z)
+type Formula interface {
+    NotX | Z
+    Y | Z
+    X | NotZ
+}
+```
+
+There are, however, two issues with this:
+
+1. A type could have *neither* of `X()` and `NotX()`.
+2. A type could have *both* of `X()` and `NotX()`.
+
+This breaks our representation, because a boolean variable always has to be
+*exactly* true or false - [it can't be neither and it can't be both](https://en.wikipedia.org/wiki/Law_of_excluded_middle).
+
+To address the first point, we define another interface:
+
+```go
+type AtLeastOne interface {
+    X | NotX
+    Y | NotY
+    Z | NotZ
+}
+```
+
+Any type satisfying `AtLeastOne` has to assign at least one of true and false
+to each variable.
+
+Similarly, we define an interface to address the second problem:
+
+```go
+type Both_X interface { X; NotX }
+type Both_Y interface { Y; NotY }
+type Both_Z interface { Z; NotZ }
+type Both interface {
+    Both_X | Both_Y | Both_Z
+}
+```
+
+Any type satisfying `Both` now assigns both true *and* false to at least one
+variable.
+
+To represent a valid, satisfying assignment, a type thus has to
+
+1. satisfy `Formula`
+2. satisfy `AtLeastOne`
+3. *not* satisfy `Both`
+
+Now, we ask our compiler to type-check this Go program[^5]:
+
+```go
+func G[T Both]() {}
+func F[T interface{ Formula; AtLeastOne }]() {
+    G[T]() // Allowed if and only if (Formula && AtLeastOne) => Both
+}
+```
+
+This program should compile, if and only if any type satisfying `Formula` and
+`AtLeastOne` also satisfies `Both`. Because we are looking at the *complement*
+of SAT[^6], we invert this, to get our final answer:
+
+```
+    !( (Formula && AtLeastOne) =>  Both )
+<=> !(!(Formula && AtLeastOne) ||  Both ) // Because A => B is equivalent to "!A || B"
+<=> !(!(Formula && AtLeastOne  && !Both)) // De Morgan's law
+<=>     Formula && AtLeastOne  && !Both   // Double negation
+```
+
+This finishes our reduction: The compiler should reject the program, if and
+only if the formula has a satisfying assignment. The Type Parameter Problem is
+at least as hard as the complement of SAT.
+
+## Going forward
+
+This is, of course, a deeply dissatisfying answer. Our `Stringish` example
```

Comment:
I think it would be good to spell out more explicitly what the above means. 

Maybe something like:

This explains why the Go spec puts this restriction in place: to avoid having to implement a SAT solver in the compiler. The restriction, of course, is deeply dissatisfying. …